File: typeclasses/Eval.hs
module Eval(Eval(..)) where

import Simple
import Rich

class Eval x where
  eval :: x -> Int

instance Eval Lit where
  eval (Lit x) = x

instance (Eval l, Eval r) => Eval (Add l r) where
  eval (Add l r) = eval l + eval r

instance (Eval l, Eval r) => Eval (Mul l r) where
  eval (Mul l r) = eval l * eval r

File: cake/src/Eval.scala
trait Eval extends Base with BaseMul {
  type exp <: Exp

  trait Exp extends super.Exp {
    def eval: Int
  }

  class Lit(value: Int) extends super.Lit(value) with Exp {
    def eval = value
  }

  class Add(left: exp, right: exp) extends super.Add(left, right) with Exp {
    def eval = left.eval + right.eval
  }

  class Mul(left: exp, right: exp) extends super.Mul(left, right) with Exp {
    def eval = left.eval * right.eval
  }
}
